src/Microsoft.ML.Data/DataLoadSave/Binary/CodecFactory.cs

```diff
@@ -54,18 +54,20 @@ public CodecFactory(IHostEnvironment env, MemoryStreamPool memPool = null)
             RegisterSimpleCodec(new UnsafeTypeCodec<ulong>(this));
             RegisterSimpleCodec(new UnsafeTypeCodec<Single>(this));
             RegisterSimpleCodec(new UnsafeTypeCodec<Double>(this));
-            RegisterSimpleCodec(new UnsafeTypeCodec<DvTimeSpan>(this));
+            RegisterSimpleCodec(new UnsafeTypeCodec<TimeSpan>(this));
             RegisterSimpleCodec(new DvTextCodec(this));
             RegisterSimpleCodec(new BoolCodec(this));
             RegisterSimpleCodec(new DateTimeCodec(this));
-            RegisterSimpleCodec(new DateTimeZoneCodec(this));
+            RegisterSimpleCodec(new DateTimeOffsetCodec(this));
             RegisterSimpleCodec(new UnsafeTypeCodec<UInt128>(this));
 
             // Register the old boolean reading codec.
             var oldBool = new OldBoolCodec(this);
             RegisterOtherCodec(oldBool.LoadName, oldBool.GetCodec);
-
             RegisterOtherCodec("VBuffer", GetVBufferCodec);
+            RegisterOtherCodec("DvDateTimeZone", new DateTimeOffsetCodec(this).GetCodec);
+            RegisterOtherCodec("DvDateTime", new DateTimeCodec(this).GetCodec);
+            RegisterOtherCodec("DvTimeSpan", new UnsafeTypeCodec<TimeSpan>(this).GetCodec);
             RegisterOtherCodec("Key", GetKeyCodec);
```

Comment:
I see the new test helped show up some issues.

BTW where are we testing the writing of data into what amounts to a new format?

Reply:
We have a test that reads from the old IDV format then it writes back as text and that is where writing of data into new format is tested. 

---
In reply to: [213349875](https://github.com/dotnet/machinelearning/pull/693#discussion_r213349875) [](ancestors = 213349875)

Reply:
Maybe a simple way to do this is modify your test a little bit.

Currently it is something like this.

```csharp
TestCore("savedata", idvPath, "loader=binary", "saver=text", textOutputPath.Arg("dout"));
```

If we modify it to be this:

```csharp
var intermediateData = /// some temp IDV file.
TestCore("savedata", idvPath, "loader=binary", "saver=text", textOutputPath.Arg("dout"));
TestCore("savedata", idvPath, "loader=binary", "saver=binary", intermediateData.ArgOnly("dout"));
TestCore("savedata", intermediateData, "loader=binary", "saver=text", textOutputPath.Arg("dout"));
```

That will test everything.